### ISSUE TITLE
High recall, low precision in my dataset

### ISSUE THREAD

**Issue by hungtran122** [open-mmlab/mmdetection#506]
Hello,

Thanks  for a neat repo.
I have trained faster_rcnn_r50_fpn_1x.py on my own dataset.
The result is unexpected. On some class, recall is very high , but precision is unbelievably low.
+-------+------+-------+--------+-----------+-------+
| class | gts  | dets  | recall | precision | ap    |
+-------+------+-------+--------+-----------+-------+
| 1     | 2982 | 15081 | 0.979  | 0.216     | 0.938 |
| 2     | 197  | 3547  | 0.919  | 0.052     | 0.164 |
| 3     | 107  | 1369  | 0.720  | 0.057     | 0.190 |
| 4     | 131  | 2420  | 0.824  | 0.045     | 0.388 |
| 5     | 0    | 17    | 0.000  | 0.000     | 0.000 |
| 6     | 150  | 6123  | 0.760  | 0.019     | 0.293 |
| 7     | 84   | 863   | 0.845  | 0.083     | 0.273 |

Hardware: 1 gpu, 1080ti
LR: 0.0025 (0.02/8.0)
Batch_size: 8

Do you have any suggestions?

**Comment by hellock** [comment]
The recall and precision is computed with all the bboxes, which is the right bottom of the PR curve. A high recall and low precision is expected.

**Comment by hungtran122** [comment]
Hello hellock,

I got your point.
My point is that precision of 1st class is improved as I train more epochs. Reversely precision of other classes are not improved at all as I train more epochs.
Do you have any suggestions? (like: hyperparameters, input size, etc.)
Appreciated

**Comment by thangvubk** [comment]
Your data is imbalanced. The first class has 2982 gts but the others have much less than that.

**Comment by hungtran122** [comment]
Hello thangvubk,

Thank for your reply.
I tried to remove 1st class and train with the rest. Same phenomenon was observed.
Do you think of any other possibilities which lead this results?

**Comment by thangvubk** [comment]
You may reduce max_dets to 200.

**Comment by hungtran122** [comment]
I'll try and back to you soon. Thanks.

**Comment by hungtran122** [comment]
Hello thangvubk,

I tried to reduce max_dets to 200.
It didn't fix the problems.

**Comment by thangvubk** [comment]
Can you share some results.

**Comment by hungtran122** [comment]
Sure. This what I got:
+-------+------+------+--------+-----------+-------+
| class | gts  | dets | recall | precision | ap    |
+-------+------+------+--------+-----------+-------+
| 1     | 2982 | 6280 | 0.979  | 0.361     | 0.956 |
| 2     | 197  | 1505 | 0.853  | 0.086     | 0.401 |
| 3     | 107  | 215  | 0.626  | 0.144     | 0.300 |
| 4     | 131  | 1356 | 0.870  | 0.042     | 0.572 |
| 5     | 0    | 110  | 0.000  | 0.000     | 0.000 |
| 6     | 150  | 865 | 0.720  | 0.068     | 0.555 |
| 7     | 84   | 153  | 0.500  | 0.165     | 0.227 |

So the aps fluctuate over training epochs (except for 1st class). It not steadily improves as expected.

**Comment by thangvubk** [comment]
Why your ``dets`` is greater than 200 when after limiting this value by 200?

**Comment by hungtran122** [comment]
max_dets is used for 1 image, right? If so, this results is for thousands images.

**Comment by thangvubk** [comment]
If so, classes 2 to 7 have ~0.1 gt per image. That why you have very low precision. 200 is still very high for your dataset. You can try to limit it to 10 x gt per image.

**Comment by hungtran122** [comment]
I got new results. Precision happens to be fixed by training on solely two classes of interest. 
However the results are still not stable. I mean even I train more at some epoch the results are good, at some others the results are bad. It seems that loss function are in a bad local optima. Would you suggest anything in this case?
+------------+-----+------+--------+-----------+-------+
| class      | gts | dets | recall | precision | ap    |
+------------+-----+------+--------+-----------+-------+
| Pedestrian | 131 | 267  | 0.863  | 0.430     | 0.770 |
| Cyclist    | 150 | 139  | 0.607  | 0.711     | 0.560 |
+------------+-----+------+--------+-----------+-------+
| mAP        |     |      |        |           | 0.665 |
+------------+-----+------+--------+-----------+-------+

**Comment by thangvubk** [comment]
It seems you have overfitting. But i dont have enough information to suggest you. You might dig deeper and try your own.

**Comment by hungtran122** [comment]
Thanks anyway. Btw. why do you think it's over-fitting. As I observe the mAP, total mAP cannot get higher than 0.68. Meanwhile at some epochs Pedestrian's ap is high and Cyclist's ap is low. At some other epochs, Cyclist's ap is high, Pedestrian's ap is low.

**Comment by Shadowalker1995** [comment]
Hi, hungtran122.
I see your output is different than me. Are you using VOC dataset or COCO dataset?
your output is VOC format, right?

**Comment by chengcchn** [comment]
@hellock I got same question. When I trained with my dataset, the results shown that the recall is high, the precision is low, but the ap and map is high enough.
![Snipaste_2019-11-14_16-26-19](https://user-images.githubusercontent.com/38886481/68839369-8cfa6700-06fb-11ea-9b17-aafa3b800bb6.png)
So is that normal for comparison with other model just considering the map? Or should I try to improve the precison?

**Comment by MuhammadAliShahzad** [comment]
I trained my network on the same dataset on another library, the average and recall remained high as expected.
When I trained it here (with default hyperparameters), the precision remained too low for some classes and high for others. Seems to be a bug, maybe.

**Comment by chengcchn** [comment]
@MuhammadAliShahzad Thank you for your reply. Can you tell me which library you used? I want to test it on my dataset too.
